doc: add TypeScript execution requirements (#44030)
nodejs/node: doc/contributing/maintaining-types-for-nodejs.md

* Automation within external type projects consumes the JSON and automatically
  generates a PR to add the API.

## High level approach - development workflow

The number of people using TypeScript with Node.js is significant enough
that providing a good developer experience is important. While TypeScript
is identified specifically, a secondary goal is that what we provide to improve

> development experience with TypeScrtipt would apply to other type

Reviewer: ```suggestion
development experience with TypeScript would apply to other type
```
Typo.

> development experience with TypeScrtipt would apply to other type

Reviewer 2: ```suggestion
development experience with TypeScript would apply to other type
```

Reviewer 3: Duplicate of https://github.com/nodejs/node/pull/44030#discussion_r932613288.

> systems as well.

Reviewer 2: ```suggestion
systems and transpiled languages as well.
```

> We have agreed that the approach will `NOT` include bundling TypeScript

Reviewer: ```suggestion
We have agreed that the approach will **NOT** include bundling TypeScript
```

> We have agreed that the approach will `NOT` include bundling TypeScript

Reviewer 2: ```suggestion
We have agreed that the approach will *not* include bundling TypeScript
```

Reviewer 3: Duplicate of https://github.com/nodejs/node/pull/44030#discussion_r932613759.

tools with Node.js but instead improve the developer experience for how
those tools are installed/configured to work with Node.js.

The high level developer experience we are working towards was captured in the
[next-10 TypeScript mini-summit](https://github.com/nodejs/next-10/pull/150)
and is as follows:

1. Users can ask Node.js to execute a file which is not one of the types it
   can execute by default (.js, .mjs, etc.). For example `node script.ts`.

> 2. On startup Node.js will look for a config which is in the scope of the

Reviewer: ```suggestion
2. On startup, Node.js will look for a config which is in the scope of the
```

   file being executed.
3. If no config is found, Node.js will echo either:
   * If the file was a TypeScript file, a TypeScript specific message with a
     reference to a link on Nodejs.org on how to install required components
     for TypeScript and how to add the associated config.

>    * If the file was not a TypeScript file, a generic message

Reviewer: ```suggestion
   * If the file was not a TypeScript file, a generic message.
```

> 1. Users can ask Node.js to execute a file which is not one of the types it
>    can execute by default (.js, .mjs, etc.). For example `node script.ts`.
> 2. On startup Node.js will look for a config which is in the scope of the
>    file being executed.
> 3. If no config is found, Node.js will echo either:
>    * If the file was a TypeScript file, a TypeScript specific message with a
>      reference to a link on Nodejs.org on how to install required components
>      for TypeScript and how to add the associated config.
>    * If the file was not a TypeScript file, a generic message
> 4. If a config is found, Node.js will extract the Node.js options for
>    that config and apply them as if they had been provided on the command
>    line.
> 5. Assuming the options associated with the config add support for the
>    type of file being loaded, Node.js will execute the file as if it was one
>    of the file types it can execute by default.

Reviewer 2: This implies that the config is specific to the unknown file, while it’s really that Node is starting up with a config that might tell Node how to handle the unknown file.

```suggestion
1. When Node.js is started with an entry point that is not a file type that
   Node.js recognizes, for example `node script.ts`, an informative error
   message is printed that directs users to a webpage where they can
   learn how to configure Node.js to support that file type.
2. Node.js gains support for loading configuration from a file. Most, if not
   all, of the configuration supported by `NODE_OPTIONS` would be
   supported in this file (which might be the `package.json` that lives
   near the entry point file). The webpage with instructions would tell
   users what configuration to put in this file to get Node.js to support
   their file type.
3. When Node.js is run with the correct configuration, either in a file or
   `NODE_OPTIONS` or flags, the unknown file type is executed as expected.
```

> Some additional specifics around the current thinking include:

Reviewer: ```suggestion
Some additional specifics around the current approach include:
```

> * loaders already provide a number of the components needed to

Reviewer: ```suggestion
* Loaders already provide a number of the components needed to
```

  satisfy the requirements above. They already provide the Node.js
  options that are needed to achieve many of the requirements above.

> * package.json as the location for the config is potentially a good

Reviewer: ```suggestion
* `package.json` as the location for the config is potentially a good
```

> * loaders already provide a number of the components needed to
>   satisfy the requirements above. They already provide the Node.js
>   options that are needed to achieve many of the requirements above.
> * package.json as the location for the config is potentially a good

Reviewer 2: ```suggestion
* Loaders already provide a number of the components needed to
  satisfy the requirements above. They already provide the Node.js
  options that are needed to achieve many of the requirements above.
* `package.json` as the location for the config is potentially a good
```

  choice as Node.js already looks for it as part of startup.

> * The implementation chosen should allow for difference configuration in
>   for different enronments/conditions like prod,dev, etc.

Reviewer: ```suggestion
* The implementation chosen should allow for different configuration
  in/for different environments/conditions like `prod`, `dev`, etc.
```

* We don't have consensus on provding an opinionated default but
  that should be explored after the initial steps are complete.
* It will be important that as part of the messaging around this
  functionality that we avoid confusion that could lead people to ship

>   TypeScript files (ex `script.ts`) instead of the processed files
>   (ex `script.js`).

Reviewer: ```suggestion
  TypeScript files (e.g. `script.ts`) instead of the processed files
  (e.g. `script.js`).
```

     reference to a link on Nodejs.org specific on learning how to 
     configure TypeScript will be provided.
   * For other file types a generic message and shared webpage will be
     used.
2. Node.js gains support for loading configuration from a file. Most, if not
   all, of the configuration supported by `NODE_OPTIONS` would be
   supported in this file (which might be the `package.json` that lives
   near the entry point file). The webpage with instructions would tell
   users what configuration to put in this file to get Node.js to support
   their file type.
3. When Node.js is run with the correct configuration, either in a file or
   `NODE_OPTIONS` or flags, the unknown file type is executed as expected. 

Some additional specifics around the current approach include:

* Loaders already provide a number of the components needed to
  satisfy the requirements above. They already provide the Node.js
  options that are needed to achieve many of the requirements above.
* `package.json` as the location for the config is potentially a good
  choice as Node.js already looks for it as part of startup.
* The implementation chosen should allow for different configuration

>   in/for different environments/conditions like `prod`, `dev`, etc.

Reviewer 4: I'm strongly -1 on adding the concept of NODE_ENV or analog to Node.js. However, the multiple environment is an issue.

This is often solved with supporting a "notable place config file" (possibly package.json) as well as a flag to specify said config file. Folks can support development with the default setting and override in prod, or viceversa.

Reviewer 2: I think it’s best for this document to leave out implementation details of how this is achieved. I think there’s consensus that we wouldn’t use `NODE_ENV`, especially because it doesn’t make much sense to mix environment variables and files to define configuration. I don’t think it’s really necessary to specifically call out that we won’t be using `NODE_ENV`.

There’s already precedent for current Node supporting prod and dev (and other) configurations via the `--conditions` flag (https://nodejs.org/api/cli.html#-ccondition—conditionscondition) with those conditions appearing in `package.json` `"exports"`. I would assume that whatever we come with either uses that or is similar to it. But I don’t think we need to get into this level of detail here. This is just defining the goal.

Reviewer 4: None of `--conditions` mentions staging, production or development.

How about:

> The implementation chosen should allow for being overridden via a different mechanism 
  for production-like environments.

Reviewer 2: > None of --conditions mentions staging, production or development.

https://nodejs.org/api/cli.html#-ccondition%E2%80%94conditionscondition mentions development. The feature was created for this use case.

"Overridden via a different mechanism" is an implementation detail. The original text is fine. We need to support different configurations for different environments/conditions. That's enough.

Reviewer 4: The original text mention specific environment names. As said, this is a deal breaker. NODE_ENV is one of the biggest mistakes we did as a community.

Reviewer 2: ```suggestion
  in/for different environments/conditions such as production
  versus development, or different types of hosted environments
  such as serverless vs traditional, etc.; Node.js would not make
  any recommendations or have any expectations as to what the
  separate configuration blocks should be named or what their
  purposes should be, just that a configuration file should have
  the ability to provide different configurations for user-defined
  conditions.
```

Reviewer 4: That's ok, thx.

* We don't have consensus on provding an opinionated default but
  that should be explored after the initial steps are complete.
* It will be important that as part of the messaging around this
  functionality that we avoid confusion that could lead people to ship
  TypeScript files (e.g. `script.ts`) instead of the processed files

>   (e.g. `script.js`).

Reviewer 4: We are missing some note on tsconfig.

I _think_ we have consensus on saying that Node.js should not dictate how TS is configured and left out to the rest of the toolchain.

Reviewer 2: I thought this was implied by the line “We have agreed that the approach will **NOT** include bundling TypeScript” but if you want an explicit mention of `tsconfig.json`, sure.

Reviewer 4: This was called out by other folks, better be precise

Reviewer 2: Can you please propose some text that would satisfy you?

Reviewer 4: ```
* There is no plan to define a default tsconfig.json for all Node.js users
```

Author: Added

Author: @mcollina added

   users what configuration to put in this file to get Node.js to support
   their file type.
3. When Node.js is run with the correct configuration, either in a file or
   `NODE_OPTIONS` or flags, the unknown file type is executed as expected. 

Some additional specifics around the current approach include:

* Loaders already provide a number of the components needed to
  satisfy the requirements above. They already provide the Node.js
  options that are needed to achieve many of the requirements above.
* `package.json` as the location for the config is potentially a good
  choice as Node.js already looks for it as part of startup.
* The implementation chosen should allow for different configuration
  in/for different environments/conditions such as production
  versus development, or different types of hosted environments
  such as serverless vs traditional, etc.; Node.js would not make
  any recommendations or have any expectations as to what the
  separate configuration blocks should be named or what their
  purposes should be, just that a configuration file should have
  the ability to provide different configurations for user-defined

>   conditions.

Author: ```
  conditions.
* There is no plan to define a default tsconfig.json for all Node.js users
```

  generates a PR to add the API.

## High level approach - development workflow

The number of people using TypeScript with Node.js is significant enough
that providing a good developer experience is important. While TypeScript
is identified specifically, a secondary goal is that what we provide to improve
development experience with TypeScript would apply to other type
systems and transpiled languages as well.

We have agreed that the approach will **NOT** include bundling TypeScript
tools with Node.js but instead improve the developer experience for how
those tools are installed/configured to work with Node.js.

The high level developer experience we are working towards was captured in the
[next-10 TypeScript mini-summit](https://github.com/nodejs/next-10/pull/150)
and is as follows:

1. When Node.js is started with an entry point that is not a file type that
   Node.js recognizes, for example `node script.ts`, an informative error
   message is printed that directs users to a webpage where they can
   learn how to configure Node.js to support that file type.
   * If the file was a TypeScript file, a TypeScript specific message with a

>      reference to a link on Nodejs.org specific on learning how to 

Author: ```suggestion
     reference to a link on Nodejs.org specific on learning how to
```

     configure TypeScript will be provided.
   * For other file types a generic message and shared webpage will be
     used.
2. Node.js gains support for loading configuration from a file. Most, if not
   all, of the configuration supported by `NODE_OPTIONS` would be
   supported in this file (which might be the `package.json` that lives
   near the entry point file). The webpage with instructions would tell
   users what configuration to put in this file to get Node.js to support
   their file type.
3. When Node.js is run with the correct configuration, either in a file or

>    `NODE_OPTIONS` or flags, the unknown file type is executed as expected. 

Author: ```suggestion
   `NODE_OPTIONS` or flags, the unknown file type is executed as expected.
```
